enhancements/cgu-precaching-additional-images.md

```diff
@@ -0,0 +1,308 @@
+
+The following changes are suggested:
+
+1. Add a new `AdditionalImages` field to the `PrecachingSpec` struct in
+`cluster-group-upgrades-operator/api/v1alpha1/clustergroupupgrade_types.go` as shown below.
+    ```golang
+    ...
+    // PrecachingSpec defines the pre-caching software spec derived from policies
+    type PrecachingSpec struct {
+        PlatformImage                string   `json:"platformImage,omitempty"`
+        OperatorsIndexes             []string `json:"operatorsIndexes,omitempty"`
+        OperatorsPackagesAndChannels []string `json:"operatorsPackagesAndChannels,omitempty"`
+        ExcludePrecachePatterns      []string `json:"excludePrecachePatterns,omitempty"`
+        AdditionalImages    []string `json:"additionalImages,omitempty"`
+    }
+   ...
+
+2. Apply the operator-sdk tool to update the TALM custom resource definitions, manifests and auto-generated
+golang source files.
+
+3. Extend the `controllers/managedClusterResources.go` and
+`cluster-group-upgrades-operator/controllers/precache.go` source files to handle reconciliation
+control logic pertaining to the `additionalImages` object.
+
+4. Extend the pre-cache template source file `controllers/templates/precache-templates.go` with the
+`additionalImages` field.
+
+5. Refactor the `cluster-group-upgrades-operator/pre-cache/precache.sh` and
+`cluster-group-upgrades-operator/pre-cache/pull` scripts to check for additional user images and pull
+them accordingly. The existing image pulling functionality should be refactored in order to promote
+code reusability and avoid code duplication.
+
+### Workflow Description
+
+The general workflow for pre-caching user specified images is as follows:
+`ClusterSelected -> Validated -> PrecacheSpecValid -> PrecachingSucceeded -> <TALM workflow continues>`.
+
+Detailed user and worklow steps are outlined below.
+
+1. User creates a CGU CR with the `preCaching` field set to `true` and prepares a ConfigMap (named
+`cluster-group-upgrade-overrides`) with the list of additional images to be pre-cached as shown in
+the examples below. Note that the ConfigMap should be created concurrently with the CGU or should
+exist prior to the creation of it. Furthermore, the ConfigMap should be created in the same namespace
+as the CGU on the hub cluster.
+   - Example CGU CR:
+    ```yaml
+    apiVersion: ran.openshift.io/v1alpha1
+    kind: ClusterGroupUpgrade
+    metadata:
+      name: cgu
+      namespace: default
+    spec:
+      clusters:
+      - sno1
+      - sno3
+      - sno7
+      enable: false
+      preCaching: true
+      backup: false
+      managedPolicies:
+      - policy0-common-config-policy
+      remediationStrategy:
+        maxConcurrency: 4
+        timeout: 240
+    ```
+   - Example ConfigMap:
+    ```yaml
+    apiVersion: v1
+    kind: ConfigMap
+    metadata:
+      name: cluster-group-upgrade-overrides
+      namespace: default
+    data:
+      additionalWorkloadImages: |
```

Comment:
After the CGU and configmap is created, can user update, i.e. add/modify/delete additional image list, if yes, do we handle those cases? Just trying to understand if that is out of scope for now

Reply:
The ConfigMap can be edited as many times as required. However, after a CGU is applied with `spec.preCaching` enabled - then a new CGU would need to be created for changes to the ConfigMap.
Furthermore, at this point in time, only additions and modifications (i.e. pre-caching different image versions) are supported. Deletions would need to be performed manually. If there is a requirement for pre-cached images to be deleted via the CGU, this can be captured as a feature enhancement in a future EP.

Reply:
I also thought it was implied by the "no UI for inspecting images" statement - I will clarify this by explicitly stating it as a non-goal.

Reply:
The `Non-goals` section has been updated to reflect the above.